google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc

```diff
@@ -0,0 +1,147 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+#include "google/cloud/storage/internal/async/rewriter_connection_tracing.h"
+#include "google/cloud/storage/mocks/mock_async_rewriter_connection.h"
+#include "google/cloud/storage/testing/canonical_errors.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/opentelemetry_options.h"
+#include "google/cloud/testing_util/async_sequencer.h"
+#include "google/cloud/testing_util/opentelemetry_matchers.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+#include <opentelemetry/trace/semantic_conventions.h>
+#include <cstdint>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::google::cloud::storage_experimental::RewriteObjectResponse;
+using ::google::cloud::storage_mocks::MockAsyncRewriterConnection;
+using ::google::cloud::testing_util::AsyncSequencer;
+using ::google::cloud::testing_util::EventNamed;
+using ::google::cloud::testing_util::InstallSpanCatcher;
+using ::google::cloud::testing_util::OTelAttribute;
+using ::google::cloud::testing_util::SpanEventAttributesAre;
+using ::google::cloud::testing_util::SpanHasEvents;
+using ::google::cloud::testing_util::SpanHasInstrumentationScope;
+using ::google::cloud::testing_util::SpanKindIsClient;
+using ::google::cloud::testing_util::SpanNamed;
+using ::google::cloud::testing_util::SpanWithStatus;
+using ::google::cloud::testing_util::StatusIs;
+using ::google::cloud::testing_util::ThereIsAnActiveSpan;
+using ::testing::IsEmpty;
+
+auto PermanentError() {
+  return StatusOr<RewriteObjectResponse>(
+      storage::testing::canonical_errors::PermanentError());
+}
+
+TEST(RewriterTracingConnection, Basic) {
+  auto span_catcher = InstallSpanCatcher();
+  AsyncSequencer<void> sequencer;
+
+  auto mock = std::make_unique<MockAsyncRewriterConnection>();
+  EXPECT_CALL(*mock, Iterate)
+      .WillOnce([&] {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
```

Comment:
I think we would also like to `EXPECT_TRUE(OTelContextCaptured())` in all of these.

Reply:
Done